pkg/resources/scaleway/resources.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scaleway
+
+import (
+	"k8s.io/kops/pkg/resources"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/scaleway"
+
+	iam "github.com/scaleway/scaleway-sdk-go/api/iam/v1alpha1"
+	"github.com/scaleway/scaleway-sdk-go/api/instance/v1"
+)
+
+const (
+	resourceTypeServer = "server"
+	resourceTypeSSHKey = "ssh-key"
+	resourceTypeVolume = "volume"
+)
+
+type listFn func(fi.Cloud, string) ([]*resources.Resource, error)
+
+func ListResources(cloud scaleway.ScwCloud, clusterInfo resources.ClusterInfo) (map[string]*resources.Resource, error) {
+	resourceTrackers := make(map[string]*resources.Resource)
+	clusterName := clusterInfo.Name
+
+	listFunctions := []listFn{
+		listServers,
+		listSSHKeys,
+		listVolumes,
+	}
+
+	for _, fn := range listFunctions {
+		rt, err := fn(cloud, clusterName)
+		if err != nil {
+			return nil, err
+		}
+		for _, t := range rt {
+			resourceTrackers[t.Type+":"+t.ID] = t
+		}
+	}
+
+	return resourceTrackers, nil
+}
+
+func listServers(cloud fi.Cloud, clusterName string) ([]*resources.Resource, error) {
+	c := cloud.(scaleway.ScwCloud)
+	servers, err := c.GetClusterServers(clusterName, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	resourceTrackers := []*resources.Resource(nil)
+	for _, server := range servers {
+		blocks := []string(nil)
+		for _, volume := range server.Volumes {
+			blocks = append(blocks, resourceTypeVolume+":"+volume.ID)
+		}
```

Comment:
Seems strange that bocks/blocked is set in both volumes and servers. Was this necessary?

Reply:
Not particularly but it seemed strange to me not to set it both way, I thought the code was more explicit/exhaustive this way but I can drop one of them if it's weird.

Reply:
Generally you would set one that it's the easiest to set, or it makes more sense to you.
In many cases, these were not set at all, assuming that kOps will retry to delete them anyway on error. This is why you don't see that many references to blocks/blocked.

Reply:
I only kept the `blocked by` field in volume as it was the easiest to set.